### PR TITLE
Ticket6079 Add Aeroflex IOC

### PR DESCRIPTION
**Pull request by Scott-James-Hurley**
Adds Aeroflex IOCs

https://github.com/ISISComputingGroup/IBEX/issues/6079

### CODE REVIEW

File: aeroflexSup/Makefile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+TOP=..
+include $(TOP)/configure/CONFIG
+#=======================================
+
+# Install .dbd and .db files
+DB += aeroflex_2023A.db
```
Comment (NikolaRoev) [inline aeroflexSup/Makefile:6]: Running `make` fails on this. The DB declarations are in lowercase while the file names are in uppercase.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    $(eval $(dir)_DEPEND_DIRS += $(filter %Sup %App, $(DIRS))))
+
+# iocBoot depends on all *App dirs
+iocBoot_DEPEND_DIRS += $(filter %App,$(DIRS))
+
+# Add any additional dependency rules here:
+
+include $(TOP)/configure/RULES_TOP
+
+ioctests:
```
Comment (NikolaRoev) [inline Makefile:36]: This should be removed if the emulator and tests are not moved.

File: aeroflexSup/AEROFLEX_common.db [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+record(ao, "$(P)CAR_FREQ_CONV")
+{
+    field(DESC, "Carr freq val converted to units")
+	field(EGU, "")
+}
+
+record(calcout, "$(P)CALC_RETURN_CAR_FREQ")
+{
+	field(DESC, "Convert car freq to kHZ,MHZ,GHZ")
+	#field(FLNK, "$(P)CARRIER_FREQ")
```
Comment (NikolaRoev) [inline aeroflexSup/AEROFLEX_common.db]: Is this comment needed or a mistake?

File: aeroflexSup/AEROFLEX_common.db [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+record(ao, "$(P)CARRIER_FREQ:SP_NO_ACTION")
+{
+	info(autosaveFields, "VAL")
+	field(PREC, "2")
+	field(UDFS, "NO_ALARM")
+}
+
+record(mbbo, "$(P)CARRIER_FREQ_UNITS")
```
Comment (NikolaRoev) [inline aeroflexSup/AEROFLEX_common.db]: This name should maybe be more descriptive. There is nothing to hint that this sets something.

**Review by NikolaRoev — CHANGES_REQUESTED** [review]
Is there a reason the emulator and tests are not in this module?
Generally, new IOCs should follow a grouped design pattern.

If the emulator and tests are just moved to the removed `system_tests` folder, everything will work as expected.
Some examples of this are `motorHuber` and  `tjmper`.

File: aeroflexSup/AEROFLEX_2030.db [hunk 5]
```diff
@@ -0,0 +1,101 @@
+record(ai, "$(P)RF_LEVEL")
+{
```
Comment (FreddieAkeroyd) [inline aeroflexSup/AEROFLEX_2030.db]: this looks the same as the one above except for PREC ? could common be loaded with a macro for PREC?

File: aeroflexSup/AEROFLEX_common.db (hunk 4 above)
Comment (FreddieAkeroyd) [reply aeroflexSup/AEROFLEX_common.db]: if it doesn't set anything, then would an `mbbi` be more appropriate?

File: aeroflexSup/AEROFLEX_2023A.db [hunk 6]
```diff
@@ -0,0 +1,112 @@
+record(ai, "$(P)RF_LEVEL")
+{
```
Comment (FreddieAkeroyd) [inline aeroflexSup/AEROFLEX_2023A.db]: if this is useful to log it should have an `info(archive, "VAL")` field

**Comment by Scott-James-Hurley** [comment]
Once all feedback has been addressed for the emulator and test PRs, I will close them without merging and add the code to this one. This will keep discussions to a single thread.

File: aeroflexSup/aeroflex_2023A.db [hunk 7]
```diff
@@ -1,35 +1,3 @@
...
-}
-
-record(ao, "$(P)RF_LEVEL:SP_NO_ACTION")
-{
-	field(PREC, "6")
-	info(autosaveFields, "VAL")
-	field(UDFS, "NO_ALARM")
-}
-
 record(mbbi, "$(P)MODE")
```
Comment (NikolaRoev) [inline aeroflexSup/aeroflex_2023A.db:1]: Are we using the separate device type DB files anywhere?

File: aeroflexSup/aeroflex_common.db [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+record(ao, "$(P)CARRIER_FREQ:SP_NO_ACTION")
+{
+	info(autosaveFields, "VAL")
+	field(PREC, "2")
+	field(UDFS, "NO_ALARM")
+}
+
+record(mbbi, "$(P)CAR_FREQ_READOUT_UNITS")
```
Comment (NikolaRoev) [inline aeroflexSup/aeroflex_common.db]:
```suggestion
record(mbbo, "$(P)CAR_FREQ_READOUT_UNITS")
```
This needs to be a `mbbo` record if we want to set it.

File: aeroflexSup/aeroflex_common.db [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+	field(UDFS, "NO_ALARM")
+	info(autosaveFields, "VAL")
+
+	field(ZRST, "HZ")
+	field(ONST, "kHZ")
+	field(TWST, "MHZ")
+	field(THST, "GHZ")
+}
+
+record(mbbi, "$(P)CARRIER_FREQ_UNITS:SP")
```
Comment (NikolaRoev) [inline aeroflexSup/aeroflex_common.db]:
```suggestion
record(mbbo, "$(P)CARRIER_FREQ_UNITS:SP")
```
Setpoint.

File: aeroflexSup/aeroflex_common.db [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+record(dfanout, "$(P)SEND_RF_LVL_PARAMS")
+{
+    field(DESC, "Set RF level button")
+    field(SCAN, "Passive")
+    field(OMSL, "closed_loop")
+    field(DOL, "$(P)RF_LEVEL:SP_NO_ACTION")
+    field(OUTA, "$(P)RF_LEVEL:SP PP")
+}
+
+record(mbbi, "$(P)MODE:SP_NO_ACTION")
```
Comment (NikolaRoev) [inline aeroflexSup/aeroflex_common.db]:
```suggestion
record(mbbo, "$(P)MODE:SP_NO_ACTION")
```
This is also a setpoint even if there is no action, we need to be able to set it.

File: aeroflexSup/aeroflex_common.db [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+    field(DTYP, "stream")
+
+    field(INP, "@aeroflex_common.proto getError $(PORT)")
+
+    field(SIML, "$(P)SIM")
+    field(SIOL, "$(P)SIM:ERROR")
+    field(SDIS, "$(P)DISABLE")
+}
+
+record(stringout, "$(P)DEV_TYPE")
```
Comment (NikolaRoev) [inline aeroflexSup/aeroflex_common.db]:
```suggestion
record(stringin, "$(P)DEV_TYPE")
```
Does this need to be a `stringout`? We just set `VAL` to the device type macro, right?

File: system_tests/common_tests/aeroflex.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+from utils.channel_access import ChannelAccess
+from utils.test_modes import TestModes
+from utils.testing import get_running_lewis_and_ioc
+from genie_python import channel_access_exceptions
+
+# Device prefix
+DEVICE_PREFIX = "AEROFLEX_01"
+EMULATOR_NAME = "aeroflex"
+
+TEST_MODES = [TestModes.DEVSIM]
```
Comment (NikolaRoev) [inline system_tests/common_tests/aeroflex.py]:
```suggestion
```
Don't really need this here.

File: system_tests/tests/aeroflex2023A.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+IOCS = [
+    {
+        'name': DEVICE_PREFIX,
+        'directory': get_default_ioc_dir('aeroflex'),
+        'macros': {'DEV_TYPE': '2023A'},
+        'emulator': EMULATOR_NAME,
+        'lewis_protocol': 'model2023A',
+    },
+]
+
```
Comment (NikolaRoev) [inline system_tests/tests/aeroflex2023A.py:23]: We do need `TEST_MODES` here though, and for the other device type test file.

**Review by NikolaRoev — CHANGES_REQUESTED** [review]
Tests fail to run. The batch file uses the IOCTestFramework so we need to add the path to common tests `aeroflex.py` to the include paths.

File: system_tests/tests/aeroflex2023A.py (hunk 13 above)
Comment (Scott-James-Hurley) [reply system_tests/tests/aeroflex2023A.py:23]: I define it in `common_tests/areoflex.py` and then import it into the other two test files. If that's an improper way to do it I'm happy to change it.

Comment (NikolaRoev) [reply system_tests/tests/aeroflex2023A.py:23]: My thinking was, we are running two different suits of tests for two different devices, down the line we might want to have different test modes for each.

**Comment by Scott-James-Hurley** [comment]
> Tests fail to run. The batch file uses the IOCTestFramework so we need to add the path to common tests `aeroflex.py` to the include paths.

How would that be done?

**Comment by NikolaRoev** [comment]
> > Tests fail to run. The batch file uses the IOCTestFramework so we need to add the path to common tests `aeroflex.py` to the include paths.
> 
> How would that be done?

```python
from utils.ioc_launcher import EPICS_TOP
sys.path.append(os.path.join(EPICS_TOP, "support", "aeroflex", "master", "system_tests", "tests", "common_tests"))
```
Something like this works.

**Review by NikolaRoev — APPROVED** [review]

**Pull request merged**